src/main/scala/analysis/RegionInjector.scala

```diff
@@ -145,16 +163,148 @@ class RegionInjector(program: Program, mmm: MemoryModelMap) {
-  override def visitMemory(node: Memory): Memory = memory
+class RegionInjectorDSA(override val program: Program, DSATopDown: mutable.Map[Procedure, Graph]) extends RegionInjector {
+  private val mergedRegions: mutable.Map[Cell, MergedRegionDSA] = mutable.Map()
+
+  private var sharedMemoryCounter = 0
+  private var stackCounter = 0
+
+  def injectRegions(): Unit = {
+    // visit reachable procedures
+    val queue = mutable.Queue[Procedure]()
+    val visited = mutable.Set[Procedure]()
+    queue.enqueue(program.mainProcedure)
+
+    while (queue.nonEmpty) {
+      val procedure = queue.dequeue()
+      for {
+        b <- procedure.blocks
+        s <- b.statements
+      } {
+        s match {
+          case store: MemoryStore =>
+            val dsg = DSATopDown(procedure)
+            val cell = dsg.find(dsg.accessIndexToSlice(store)).cell
+            val mergedRegion = if (mergedRegions.contains(cell)) {
+              mergedRegions(cell)
+            } else {
+              val region = createRegion(cell)
+              mergedRegions(cell) = region
+              region
+            }
+            val newMemory = replaceMemory(store.mem, mergedRegion)
+            store.mem = newMemory
+          case load: MemoryLoad =>
+            val dsg = DSATopDown(procedure)
+            val cell = dsg.find(dsg.accessIndexToSlice(load)).cell
+            val mergedRegion = if (mergedRegions.contains(cell)) {
+              mergedRegions(cell)
+            } else {
+              val region = createRegion(cell)
+              mergedRegions(cell) = region
+              region
+            }
+            val newMemory = replaceMemory(load.mem, mergedRegion)
+            load.mem = newMemory
+          case _ =>
+        }
+      }
+      visited.add(procedure)
+      for (call <- procedure.calls) {
+        if (!queue.contains(call) && !visited.contains(call)) {
+          queue.enqueue(call)
+        }
+      }
+    }
+    transformMemorySections(program)
+  }
+
+  private def createRegion(cell: Cell): MergedRegionDSA = {
+    val stack = cell.node.get.flags.stack
+    val name = if (stack) {
+      stackCounter += 1
+      s"stack$$${stackCounter}"
+    } else {
+      sharedMemoryCounter += 1
+      s"mem$$${sharedMemoryCounter}"
+    }
+    MergedRegionDSA(name, cell, stack)
+  }
+
+  private def replaceMemory(memory: Memory, mergedRegion: MergedRegionDSA): Memory = {
+    if (mergedRegion.stack) {
+      StackMemory(mergedRegion.name, memory.addressSize, memory.valueSize)
+    } else {
+      SharedMemory(mergedRegion.name, memory.addressSize, memory.valueSize)
+    }
+  }
+
+  private def transformMemorySections(program: Program): Unit = {
+    // if any addressRange entries point to a mergedRegion: grab the bytes
+
+    val dsg = DSATopDown(program.mainProcedure)
+
+    // need to check if nodes match too
+
+    dsg.globalMapping.foreach { (range, field) =>
+      val cell = dsg.find(field.node.cells(0))
```

Comment:
field.node.getCell(field.offset), value of field offset may have changed in bottom up or top town phase

Reply:
How do I correctly get the cell that a globalMapping entry corresponds to then?

Reply:
dsg.find(field.node.getCell(field.offset)) should give you the cell. 

The actual offset which field.offset corresponds to in the node returned by dsg.find is slightly tricker. You can call dsg.find(field.node) which returns a field. The corresponding offset in the new node is
 field.offset + dsg.find(field.node).offset

Reply:
You are correct that dsg.find(field.node.getCell(field.offset))  returns a cell. 

But we also track slices which are internalOffsets into the cells. 

If we have 
cell = dsg.find(field.node.getCell(field.offset))
cellStart= cell.offset
cellEnd = cell.offset + cell.largestAccessedSize

then 
field.offset + dsg.find(field.node).offset
will give you a number between cellStart and cellEnd

Reply:
Ah ok, I don't think I need to care about the internalOffsets here so dsg.find(field.node.getCell(field.offset)) should work?

Reply:
Yeah it should give you the right cell